### ISSUE TITLE
Discussion on the future of Phobos

### ISSUE THREAD

**Issue by Metadorius** [Phobos-developers/Phobos#1458]
I think the time is due for me as a head of this project to discuss something important for all of us. Even if I don't directly contribute or review code at the moment, it is still important to talk about this.

Key problems:
- We have an influx of contributors with contributions of varying quality, but not enough mature developers to review the contributions and decide whether they are up to our standards, help contributors learn and grow;
- There's not enough motivation for developers to keep improving and maintaining Phobos *as a project* (this is mostly because we are not sustainable and can't be, we can't pay salaries because we can't earn money, hence people only work as long as it's interesting enough to do so);
- We have no development direction apart from "improve the game" - so far we are randomly fixing and implementing stuff without defined plans (and scope).

The result? You see it:

![image](https://github.com/user-attachments/assets/906ce14b-a8e3-45bc-bb01-080700986882)

Many of those are in the "fridge" for multiple years already. On top of that, almost all of our maintainers are mostly gone, and the development stalls further and further. For me personally it's sad to see, because:
- Phobos is my first (and so far the only) serious project that I achieved something with, and I don't want to see it go (even if I personally can't actively handle it atm)
- Phobos was intended to be *the* community project to which people will be able to contribute to, and so far this doesn't seem to be fulfilled with the contributions being left in fridge.

All of the listed problems are not something you can easily solve, and I personally don't have an answer to how to solve them at the moment, therefore, I invite any and all contributors and whoever has something to say on the topic to discuss this.

@Starkku @chaserli @MortonPL @ZivDero @FS-21 @tomsons26 @CCHyper @CrimRecya @TaranDahl @NetsuNegi @Multfinite @handama @DeathFishAtEase @Aephiex @Rampastring @Otamaa @secsome and whoever else

**Comment by DeathFishAtEase** [comment]
In fact, due to the lengthy cycle of PR merges for this project, many contributors eventually maintain their own branches and continue to develop them independently, which results in an extremely difficult merging process.
And both the project itself and the branches maintained by contributors have ultimately ended up being tailored to serve specific projects or a few projects. For instance, many features in the current version of this project are essentially functionalities that Starkku has developed for his own mods and shared. 
Many users come forward with demands that others may not necessarily need, yet highly customized logic could be delegated to engines like DynamicPatcher, rather than engines like Ares and Phobos.
However, it's clear that this project cannot be completely redone, and even engines that offer functional components for users to assemble, like the KratosPP project, already exist.

Moreover, for the current Phobos, those PRs still need to be addressed, which requires some method for prioritization and thus necessitates a 'vote' from users, even in the form of reacting with emojis (just an example, which is obviously not conducive to statistics). Addressing a portion of the PRs does allow the project to move forward, but of course, this does not fundamentally solve the problem.

**Comment by Metadorius** [comment]
> Moreover, for the current Phobos, those PRs still need to be addressed, which requires some method for prioritization and thus necessitates a 'vote' from users, even in the form of reacting with emojis (just an example, which is obviously not conducive to statistics).

This is already there in discussions, one of the reason I moved all suggestions to discussions, but for some reason no one uses the upvotes:

![tumbleweed](https://github.com/user-attachments/assets/78e13fd3-df31-4ba0-bdf5-50f5eaab48f7)

**Comment by Metadorius** [comment]
> That is to say, there is an issue where 'Many users come forward with demands that others may not necessarily need' and submit these demands to this project, but these highly customized demands can be entirely handed over to engines like DynamicPatcher, rather than Ares and Phobos.

This is one of the reasons why I am usually pushing users to build generalized features. We must provide building blocks so people build their own stuff, not very specific features.

**Comment by DeathFishAtEase** [comment]
> > That is to say, there is an issue where 'Many users come forward with demands that others may not necessarily need' and submit these demands to this project, but these highly customized demands can be entirely handed over to engines like DynamicPatcher, rather than Ares and Phobos.
> 
> This is one of the reasons why I am usually pushing users to build generalized features. We must provide building blocks so people build their own stuff, not very specific features.

But there are already KratosPP projects that have taken the lead in this direction, and it is very difficult to design something that is different yet equally universal. Users who can carry out such functional design often have a knowledge of engine functionality that is no less than that of engine developers (although they are often the same group of people)
However, even the mechanism of providing modules for users is not widely accepted by many users. The vast majority of users clearly prefer a single function to directly meet their needs. Even when I talk to other users about the KratosPP project or YR engine's use of multiple functions to achieve results, many people tend to hastily end with the phrase "the learning cost is too high, I don't have the energy, I don't want to learn". After all, highly customized logic is much more convenient for specific projects than splicing different logical modules. Of course, this is not easy to comment on for the development of a community public engine

**Comment by secsome** [comment]
The three key problems have revealed the most important problem already. I'd like to give some of my opinion about these points below.
# Key problems
## The code quality
We do have a coding-standard, but haven't enforce people do that yet. The code quality can usually be separated into several parts: 
- Does it obey the code naming standard?
- Does the new code have side effect on the old codes?
- Does the new code have been implemented in the proper way?

The naming rule is the basic one. But sadly, in the latest `develop`, there are too many places do not follow the rules. So my first suggestion is to add a step in our PR workflow. But to do this, a complete check on the project is required, and we may need to make an agreement about the naming rule. Those would take at least a week to be done. This is our basic coding-style check, and it will force the PR meet all requirements. 

Another big problem is sometimes the new code is potential inefficiency, as you can always implement this feature in a better way.  The main problem for this is this is a community based project and we don't have some plans, or iterations which lead our goal. So the main contributors cannot tell the newcomers whether there's a better way to implement the feature before the newcomer make the PR, and as it's a community based project, many people work at the time they are free, when they just want to do that, and their motivation fades really quickly, then this PR would never be completed!

As I've mentioned above, Phobos is a community based project which has really few active main contributors now. Usually the main contributor would also charge in make sure the new PR have no side effect to the existing things. So testing people are needed, while we can hardly find any of them and we cannot test all cases currently.

## Lack of motivation for being a maintainer
As @Metadorius mentioned at the beginning, we don't earn money from this project, and for most developers, at least for me, take contributing to the project as a hobby to do at my spare time. I contribute to this project because I want to give my power to the community, but this isn't my work, so I can only have really few time spend on this. (And after coding from 10 a.m to 6 p.m everyday, I don't want to more code when I get back to home honestly, I'd rather to something else). But an active maintainer is necessary for any project, whether it's driven by the community or not. 

## No clear objectives
This is something between the two points above, as we don't have an active maintainer, we cannot make plans to achieve, such as fix any vanilla bugs, implement any new features or resolve existing Phobos bugs. And it also made us hard to keep the code quality on a high level.

# Solution?
My optinion is an/some active maintainers is necessary. Everything I said below requires a/many active maintainer(s). 

I suggest we should make some plans, such as a release version every two months or some other intervals, and maintainers can arrange the things from the issues/existing pull requests at the beginning of this *release cycle*. And before the *release cycle* ends, we can give an experience pack to the modders, which is a time buffer for us to resolve the issues before the release version. And I suggest we to create branches for them with the naming like release/MAJOR.MINOR.BUILD.HOTFIX. So we can move to work on the next release cycle.

And if this is implemented, we can disallow people to push to branch develop directly. The hotfix should be pushed to the corresponding release/x.x.xx.x branch and develop only accepts the PRs. This would also help in management.

A true problem ahead is that as Phobos is growing HEAVIER and HEAVIER, modders begin to doubt if they really need this/that feature. I don't want to use someone's feature, but as it has been added into the latest version of Phobos I have to use it, even it will cause my FPS drops. So I have to stop upgrading Phobos, seeking for other self modified version, which is really bad for the consistent developing for Phobos itself. So by seperating all project into separate features, we can give the modder the ability to control which features should be enabled and which are not. We already have many feature have the switch, but the maintainer should take this reponsibility to decide which feature/bugfix need it.

In my opinion, an active maintainer is the key for Phobos to go further. And yeah I don't have any good ideas about that. But I think this is just where the key problem lies. Our discussion should focus on *how to find an active maintainer* and *how to make the system work better with as least active mainainer* I guess.

**Comment by Aephiex** [comment]
### I. Help, unmerged PRs are stacking!

That sounds like the project is lack of active collaborators who have write access. And my suggestion is simple, just recruit some voluntary collaborators from active, responsible, and mindful contributors, and allow them merge whatever look good without too much discussion, and try to trust them to not show some destructive behavior.

About myself, I think I'm doing my best as who I am, or who I think I am. I make whatever I want, I intensively test my stuff, and sometimes help people doing something. That's the best someone without write access can do, because I can do nothing in advance. Eventually someone with write access will be needed, because a castle 99% built is a castle not built. I never asked, for whatever reason, such as stability regards, or simply not having the time to, it is always the collaborators' rights to decide how to spend their free time and whether to merge or even to look at the pull requests, and asking why you don't spend more of your free time on a non-commercial project doesn't sound like a polite thing to do. Can't blame people starting to maintain their own branch, even I do, and I will definitely use my play branch as a basis if someday I will ever create a mod for myself.

Don't forget, that RA2 is an old game that you can't expect many new comers, not to say new developers, and we who are still enthusiastic modding it (or used to be) are more likely to be more and more occupied by non-hobby activities such as career, marrage, children, or as simple as new fancy games like Black Myth Wukong. That's to say, maintaining a high standard can lead to a strictly limited human resource pool, among an already selective few group who can develop, among an already shrinking fanbase of a sunset title.

_(If you really want to take my advice of recruiting new collaborators then don't come to me, I'm not a taker of long-term responsibility unless it's my job.)_

### II. Lego system?

Regarding KratosPP, in my narrow sights, the learning cost surpasses even c++ itself, and the resulting rulesmd.ini code are not straightforward and hard for another person to read how does it work out, that's not my prefered way to mod a game, and that's why I won't use it and would rather c++ instead. It might have a lower learning cost than c++ to some, but is still difficult to use, as it usually asks for ~20 lines of rulesmd.ini code to do something you can describe in one sentense with English.

I agree that the features should be generalized and as customizable as possible, but I don't really want to make a lego system, that would take years to design and to develop, and take weeks for a modder to learn how to use it, if I had that much free time I get more interesting things to do, such as playing new games, watching anime, and sampling decicious food, whatever it is, it is more fun than developing a lego system for a game like RA2, as in my sights, c++ itself is already a lego system, just you have to make your pieces from crude oil.

The closest thing in my todo list is the event handler system that user Metadorius (Kerbiter) mentioned in the issues panel. Can't believe whatever comes out from my hands will be something as lego-ish as KratosPP is, since I want to make it straightforward and keep the rulesmd.ini code readable.

### III. Objectives?

Don't really think we need that. Some of our shinest creations, like unit tooltips and the super weapon sidebar, each may already present in a newer CNC game, and we are replicating these features on a game too old to have such on the launch. A modder can simply choose to mod on a newer CNC game and they start with many features that would take us months to develop and decades to merge, and even those they don't start with they can make a short work of it much faster than we can. That's said, we are still here because we like RA2 or want to stay here for whatever reason. Don't forget, that modding is for fun. Implementing features is fun on its own, but taking a long term objective is not for me.

### Conclusion

Thank you for reading.

**Comment by Metadorius** [comment]
@secsome 

About the code quality: so far I am yet to see the overlooked bad naming to cause some kind of a fault. Yes we do stray away from code style here and there, but I think it's one of the minor-est issues.

For stuff like that - there is clang-tidy and clang-format. They need to be intergrated into our workflows in CI and in the project. They are supported by all IDEs and they do provide automatic code style checks and linting. @chaserli was working on it but didn't finish it.

> In my opinion, an active maintainer is the key for Phobos to go further.

The main problem: where should we find one?
1. The person needs to be well versed in software engineering.
2. The person needs to understand (ideally be interested in) Yuri's Revenge.
3. The person needs to be passionate to basically work on the project for free, especially tackle the usually not interesting stuff like reviewing and housekeeping.

This is basically trying to find a unicorn. Such a person is very rare.

**Comment by Aephiex** [comment]
> @Aephiex
> 
> on Lego system: I don't think we need to go as far as Kratos goes, but there is still something to improve.
> 
> 1. The event handler thingie in it's most basic form (I don't think the stuff with filters, multipliers, custom event invocation is needed for the first iteration).
> 2. Making every attachable thing (like shields, AE, possibly Laser Trails, Techno Attachments in future) all follow the same logic and managed by the same code. This will save us from reimplementing attach/detach stuff every time. Basically for the end user it will be almost the same INI code but more versatile and generic.
>    This is also known as Entity-Component pattern (NOT Entity-Component-System! two entirely different things). Every modern engine uses this pattern to allow for generic behavior. It's the building blocks of type "Want movement? Add `MovementComponent`", not the "learn a cryptic language of fake weapons, units, superweapons".

That's ideal for software engineering and when making an indie game you have absolute control over the whole project. I know Unity engine work that way and I like it alot, it makes modding of Unity games easy to do. Don't forget we don't have control over the codes of Yuri's Revenge and Ares engine extension, it is not easy to do even as simple as to fix a bug with them, and it is the nature of c++ programs, that most of the info that will help you understand will be lost after it was compiled, and the program already present are completely not following the entity-component pattern. Making something like that on top of such complexity is even more consuming than it sounds like.

**Comment by Metadorius** [comment]
> That's ideal for software engineering and when making an indie game you have absolute control over the whole project. I know Unity engine work that way and I like it alot, it makes modding of Unity games easy to do. Don't forget we don't have control over the codes of Yuri's Revenge and Ares engine extension, it is not easy to do even as simple as to fix a bug with them, and it is the nature of c++ programs, that most of the info that will help you understand will be lost after it was compiled, and the program already present are completely not following the entity-component pattern. Making something like that on top of such complexity is even more consuming than it sounds like.

Yeah absolutely. It should be an inspiration, not a direct copy.

Entity-Component pattern is mostly based on OOP principles, not C#-specific reflection and such, so I think the core idea can be implemented so it is convenient for both us and users.

**Comment by DeathFishAtEase** [comment]
> The three key problems have revealed the most important problem already. I'd like to give some of my opinion about these points below.
> 
> # Key problems
> ## The code quality
> We do have a coding-standard, but haven't enforce people do that yet. The code quality can usually be separated into several parts:
> 
> * Does it obey the code naming standard?
> * Does the new code have side effect on the old codes?
> * Does the new code have been implemented in the proper way?
> 
> The naming rule is the basic one. But sadly, in the latest `develop`, there are too many places do not follow the rules. So my first suggestion is to add a step in our PR workflow. But to do this, a complete check on the project is required, and we may need to make an agreement about the naming rule. Those would take at least a week to be done. This is our basic coding-style check, and it will force the PR meet all requirements.
> 
> Another big problem is sometimes the new code is potential inefficiency, as you can always implement this feature in a better way. The main problem for this is this is a community based project and we don't have some plans, or iterations which lead our goal. So the main contributors cannot tell the newcomers whether there's a better way to implement the feature before the newcomer make the PR, and as it's a community based project, many people work at the time they are free, when they just want to do that, and their motivation fades really quickly, then this PR would never be completed!
> 
> As I've mentioned above, Phobos is a community based project which has really few active main contributors now. Usually the main contributor would also charge in make sure the new PR have no side effect to the existing things. So testing people are needed, while we can hardly find any of them and we cannot test all cases currently.
> 
> ## Lack of motivation for being a maintainer
> As @Metadorius mentioned at the beginning, we don't earn money from this project, and for most developers, at least for me, take contributing to the project as a hobby to do at my spare time. I contribute to this project because I want to give my power to the community, but this isn't my work, so I can only have really few time spend on this. (And after coding from 10 a.m to 6 p.m everyday, I don't want to more code when I get back to home honestly, I'd rather to something else). But an active maintainer is necessary for any project, whether it's driven by the community or not.
> 
> ## No clear objectives
> This is something between the two points above, as we don't have an active maintainer, we cannot make plans to achieve, such as fix any vanilla bugs, implement any new features or resolve existing Phobos bugs. And it also made us hard to keep the code quality on a high level.
> 
> # Solution?
> My optinion is an/some active maintainers is necessary. Everything I said below requires a/many active maintainer(s).
> 
> I suggest we should make some plans, such as a release version every two months or some other intervals, and maintainers can arrange the things from the issues/existing pull requests at the beginning of this _release cycle_. And before the _release cycle_ ends, we can give an experience pack to the modders, which is a time buffer for us to resolve the issues before the release version. And I suggest we to create branches for them with the naming like release/MAJOR.MINOR.BUILD.HOTFIX. So we can move to work on the next release cycle.
> 
> And if this is implemented, we can disallow people to push to branch develop directly. The hotfix should be pushed to the corresponding release/x.x.xx.x branch and develop only accepts the PRs. This would also help in management.
> 
> A true problem ahead is that as Phobos is growing HEAVIER and HEAVIER, modders begin to doubt if they really need this/that feature. I don't want to use someone's feature, but as it has been added into the latest version of Phobos I have to use it, even it will cause my FPS drops. So I have to stop upgrading Phobos, seeking for other self modified version, which is really bad for the consistent developing for Phobos itself. So by seperating all project into separate features, we can give the modder the ability to control which features should be enabled and which are not. We already have many feature have the switch, but the maintainer should take this reponsibility to decide which feature/bugfix need it.
> 
> In my opinion, an active maintainer is the key for Phobos to go further. And yeah I don't have any good ideas about that. But I think this is just where the key problem lies. Our discussion should focus on _how to find an active maintainer_ and _how to make the system work better with as least active mainainer_ I guess.

Regarding the implementation of standards:
The implementation of this standard itself has some difficulties due to the original PR merge cycle issues, and there are many PRs that have not been merged for a long time. Obviously, it is not convenient for you to renovate the entire functionality on behalf of the PR submitters who may have lost contact during the standardization process, and there are also PRs with the same functionality submitted by different contributors.

Regarding the release cycle:
The Phobos project's build version release interval is already close to the length you mentioned, but it seems to have little effect, and there are even cases where many users no longer follow updates due to issues like build34 forcibly enabling $Include. Also, can you describe your idea in more detail? Do you mean to release a new version and then have other contributors provide patches to complete the release version?

Regarding whether users feel they need a certain feature:
In fact, I had once conceived an idea for such a tool many years ago: to use an idiot-proof, end-user facing merge tool to generate the dlls needed by users from the source code, where users can freely select the features they want in the tool (for this project, it can be extended to those PRs). If a feature/PR causes problems, users can directly trace back to the feature developer/PR contributor or even the contributor of the conflicting PR, for coordination. But I was already questioning at that time whether such a tool could be made, or how it could be accomplished. As you mentioned, regarding the current state of the code standards issue, end-users often lack the ability to handle merge conflicts. And as the source code updates, the tool itself also needs to be upgraded, and users often need to follow the version of the tool, which means it is not universal. And it is also difficult for the tool to be very intelligent and complex - complexity may not be convenient... There are many problems, obviously, such a thing is not easy to design and produce.

**Comment by DeathFishAtEase** [comment]
> @secsome
> 
> About the code quality: so far I am yet to see the overlooked bad naming to cause some kind of a fault. Yes we do stray away from code style here and there, but I think it's one of the minor-est issues.
> 
> For stuff like that - there is clang-tidy and clang-format. They need to be intergrated into our workflows in CI and in the project. They are supported by all IDEs and they do provide automatic code style checks and linting. @chaserli was working on it but didn't finish it.
> 
> > In my opinion, an active maintainer is the key for Phobos to go further.
> 
> The main problem: where should we find one?
> 
> 1. The person needs to be well versed in software engineering.
> 2. The person needs to understand (ideally be interested in) Yuri's Revenge.
> 3. The person needs to be passionate to basically work on the project for free, especially tackle the usually not interesting stuff like reviewing and housekeeping.
> 
> This is basically trying to find a unicorn. Such a person is very rare.

And most importantly, he is willing to take over this project :P

**Comment by Metadorius** [comment]
> there are even cases where many users no longer follow updates due to issues like build34 forcibly enabling $Include

Development builds were always reserving the right to be a testing ground. I don't think we ever advertised dev builds as something stable and robust, so I don't know where the expectations of those users come from.

We try to do our stuff reasonably, and we expect people staying with us to be also reasonable. It is not our job to cater to false expectations or persuade users who don't understand or even hate the project, so I don't think this is a problem.

**Comment by DeathFishAtEase** [comment]
> so I don't know where the expectations of those users come from.
> 
It is indeed the case that they are just test versions, but those test versions have indeed had a negative impact. Moreover, I've found that many users perhaps because the release version is far behind the development version or for other reasons also regard the Build versions as sufficiently 'official' and place high expectations on them, and it seems difficult to change this concept. xD

**Comment by Aephiex** [comment]
### IV. Quality issues

If something confirmed to be working can somehow break for an update, how can I feel safe grab the latest update as someone who don't code? I have encountered 2 incidence where a certain Phobos feature or fix caused an unexpected side effect.
1. The AI parallel queue customization is described to allow the similar feature in Ares to take precedence when the latter is enabled, but it is a lie, it in fact renders the latter not working.
2. The "Building upgrades now consistently use building's `PowerUpN` animation settings..." is known to make the Foehn barracks (Mental Omega) not look right, it always look like it had the knightfall extension even if it in fact had the netrunner protocal extension. Commenting out relative codes fixed it.

Having an unforseen side effect in the code is something inevidable in software engineering, sometimes old stuff can break by stupid reasons, not to mention an interest driven project that not all of us are trained to be careful on such regards. I know it is hard to ask for people to take more responsibilities, but please at the very least make sure every line in your doc is confirmed, not supposed, and if something can have a side effect, give an option to turn it off. I'm following these in my pull requests.

Lower standards, quality issues. Higher standards, everyone gets a custom branch. It is really hard to find a balance, when we are absolutely lack in active people.

**Comment by Aephiex** [comment]
> It is indeed the case that they are just test versions, but those test versions have indeed had a negative impact. Moreover, I've found that many users perhaps because the release version is far behind the development version or for other reasons also regard the Build versions as sufficiently 'official' and place high expectations on them, and it seems difficult to change this concept. xD

That's an issue caused by low level of activity and low frequency of stable release update. Despite such, even the latest dev build isn't new enough, that many useful features are still lying in the pull request panel, they aren't even becoming part of a dev build.

I SEE NIGHTLY BUILDS ON MOD RELEASES CAN YOU BELIEVE ME???

**Comment by Metadorius** [comment]
Note to myself: it appears that the current release schedule/versioning scheme is suboptimal and can be made simpler to follow, with more frequent releases and less burden on the maintainers (while also adding a possibility of better automation of stuff like changelog generation).

**Comment by Metadorius** [comment]
Few more contributors who I forgot to mention: @Fryone @ststl-s @Super-StarX @Belonit @Coronia

**Comment by Coronia** [comment]
Here're some personal takes, might not be correct:

**ABOUT CODE QUALITY**
I know that Phobos (and to an extent, many community project) suffers from coding standard issues, and if it's an other project with more popularity I would definitely support doing more standardization work. But that's not the case now cuz RA2 is an old game with decreasing popularity by itself, and modifying its engine already requires knowledges that many people don't know. We're already the 'selected few' who'd still like to working on it, and there's no salary and such to inspire people. The only 'award' is the sense of achievement: someone makes a function and sees it works, then it's been accepted by the public, so he has the motivation to contribute next time. I've seen some people don't really have much experience in coding before, but still trying their best to learn and make contribution, only because they love the game and still have interest and passion in modding.

But in this case, coding standardization might discourage such passion cuz again, it requires further ability for participants and reduces the chance of merging pull request naturally, which is also a reason that there're many PRs being left there: once they losing the passion to continue, or having real life business, they won't be bother to do further maintaineance on it. Or instead, they'll develop their own branch instead since it doesn't have that much rules to follow.

Again, I'm not against code quality by itself, but pushing it might result in ineffectiveness and de-motivation (considering people already don't have much motivation now). So a balance should be made, but so fat I don't have much idea on it either. The only thing I can think about as of now is: If a pull request just has some minor coding style problems, then maybe there's no need to ask the contributor to modify it by themselves. Maintainers could do it when merging instead as it's only a few minutes work, and they have the writing access.

**ABOUT MAINTAINER**
Here's an opinion due to my own working experience: The one who maintains the project and decides on the direction of development **MIGHT NOT** have to be a programmer. In an IT company there often be 3 types of characters: Developers, testers and product designers (or PD for short), and PDs are the one who really decides where the project should go. They don't need to do coding by itself or have any knowledge about coding, but they need to know what's really needed for this project, and have the ability to organize others to work on it.

For the case of Phobos: in before there're criticisms about it 'doesn't know what's a modder truly want', considering some of our functions are impractical in actual modding scenarios. And as of now we also don't have much goal but to make the game better, but again the definition of 'better' is varied by people. Considering this 2 things, maybe it's better that we inviting some real modders to help with the maintaineance. Phobos can't standalone with a YR mod after all, so let's say modders are the 'clients' of this project, and we as a 'provider' is to provide them with what they need. Again, they don't have to know about coding, but could just give instruction such as 'what functions are still needed' and 'which pull request should be merged first',then the other programer help with execution.

**ABOUT REDUNDANT FUNCTIONS**
TBH, I think this is due to the nature of Phobos being a community project. Again, RA2 engine extension is already a group with very small amount of participants, and it can't shine alone without an actual YR mod. Despite we're trying to make our functions as universal as we could, there're still many explict requirements from different mods that can't always be satisfied by a public build. But if we're adding too much of these custom functions and additions in the official build, it'll create extra performance tax and might even break existing stuffs unintendedly, which is bad for those who don't need these functions. As of now both those who don't get their desired functions and those who get interrupted by new additions might adbandon official build and make their own branches instead.

I think we should allow users to decide which functions they want to use, and cut out those they don't really need manually in order to prevent them from disruption their mods. 2 methods to achieve that:

Simpler approach: adding toggles for several functions, especially those that needs to be calculated per frame. For example:
```ini
[Phobos]
PassengerDeletion=false
```
If this is set, then everything about PassengerDeletion will be omitted, saving the cost of its calculation.

Harder approach: we make a branch with only basic framework, and split all functions into standalone pull requests. If someone only want to make use of a few functions in it, he can make a build with only those included. In this case, instead of a unified official build, people could assemble their own build in a more modular way. But it also has the disadvantage of requiring more work and knowledge in the field of coding.

**Comment by Fryone** [comment]
Small note. I'm far from being an expert, so I don't know what Phobos should be in total. But at least, it's good stuff to expand YR engine, so I don't see why it should be super-customizable, cause there is OpenRA for that. Of course, a lot of similar logic stuff should be united and generalized, but sometimes it's hard to find borders, cause everything seems to be connected in some sort. 

About big PR number, there is a lot of small PRs with fixes or small stuff plus some PRs for ideas/discussions/demos, so it's not that bad.

**Comment by hejiajun107** [comment]
In my opinion, It's unnecessary to worry about tons of the pull-requests leave here.
Ensuring the stability of the current version is more important.
Since phobos is an open-source project ,anyone those who wants new features in pr can maintaining and relaese their own branch under the license.
On the other hand, that would be better if a plugin structure could be designed.The authors of the plugins may responsible for maintaining themselves, and users can choose which plugin they need.

**Comment by TaranDahl** [comment]
我认为可能的解决方案：I think possible solutions are:

> We have an influx of contributors with contributions of varying quality, but not enough mature developers to review the contributions and decide whether they are up to our standards, help contributors learn and grow;

降低代码审核的标准。审核只需确保：
Lower the standards for code review. The review only needs to ensure:
1. 最基本的可读性。变量名、函数名不能是“aaa”之类的东西。大部分代码意义明确，不明确的部分要求贡献者补充注释。
考虑到我们目前的人力资源，我不认为花时间在代码风格等细枝末节上是明智的行为。另外，不要忘了我们是一个基于gamemd逆向成果构建的项目——我们的开发者们都能看懂IDA输出的那些反编译代码了，区区代码风格真的会影响到他们吗？
The most basic readability. Variable names and function names cannot be things like "aaa". Most of the code has clear meaning. For the unclear parts, contributors are required to add comments. 
Considering our current human resources, I don't think it is wise to spend time on minor details such as code style. Also, don't forget that we are a project built based on the reverse engineering results of gamemd - our developers can understand the decompiled code output by IDA. Will mere code style really affect them?
2. 除了bug修复类pr，在不开启任何pr功能的默认情况下，已有的东西（gamemd的、Ares的、Phobos已有的其它功能）应当以原有的方式运作。
pr本身的功能能否运转，通常来说pr作者比审核者要更清楚，我们可以假设它们在大多数情况下工作良好。而且正如上面多次被提及的，很多提交的功能缺少通用性，这意味着当我们考虑接收一个新功能时，可能大部分使用者都**不会**用到这个功能。所以只要我们确保它不会影响已有的东西，那么我们就满足了大部分使用者的需求。
Except for bug fix pull requests, by default without enabling any pull request functions, existing things (those of gamemd, Ares, and other existing functions of Phobos) should operate in the original way.
Whether the function of the pull request itself can operate is usually clearer to the pull request author than the reviewer. We can assume that they work well in most cases. And as mentioned many times above, many submitted functions lack universality, which means that when we consider accepting a new function, probably most users will **not** use this function. So as long as we ensure that it does not affect existing things, then we meet the needs of most users.

> There's not enough motivation for developers to keep improving and maintaining Phobos as a project (this is mostly because we are not sustainable and can't be, we can't pay salaries because we can't earn money, hence people only work as long as it's interesting enough to do so);

解决办法同上。如果你没法开源，那么你就得节流。把宝贵的人力资源省着用，用在那些对大多数人来说有意义的地方。
The solution is the same as above. If you can't increase the income, then you have to reduce the outcome. Save and use precious human resources sparingly and apply them to places that are meaningful to most people.

> We have no development direction apart from "improve the game" - so far we are randomly fixing and implementing stuff without defined plans (and scope).

我不认为这是个需要解决的问题。尤里的复仇引擎就像一台型号老旧而且本来就粗制滥造的机器，把西木留下的这台破烂修好本来就是这个游戏最需要的开发方向之一。
I don't think this is even a problem that needs to be solved. The Yuri's Revenge engine is like an old and shoddily made machine. Fixing this junk left by Westwood is one of the most needed development directions for this game.

**Comment by DeathFishAtEase** [comment]
> On the other hand, that would be better if a plugin structure could be designed.

This is why I said we need 'an idiot-proof, end-user facing merge tool,' but I believe creating such a tool is equally challenging. 
> The authors of the plugins may responsible for maintaining themselves, and users can choose which plugin they need.

But I think the feasibility of creating a similar tool for developers is relatively high, as you mentioned here.

**Comment by Metadorius** [comment]
> Also please stop with those force push abuses, why not discuss it before get merged instead nuking people stuffs that already mereged , jeez .

I also think it is not correct to handle stuff like this and I am already talking with people about this. Apologies for this.

**Comment by Super-StarX** [comment]
Phobos is still a powerful drive for the community, continuing to push its development. However, due to the shortage of programmers and lack of financial support, several issues have arisen.  
For developers, the main challenge lies in the lack of developers to maintain PRs and fix bugs. For users, the issues include bugs and a large number of "for self fun" features created by individual developers, leading to a very unstable experience.  

I believe the future of the engine extension lies in **plugin-based** or **custom toggle-based** approaches:  
1. **Plugin-Based**: Each feature or organization can develop a standalone DLL, and a mod can utilize multiple DLLs.  
   - We have already made some preliminary attempts in this direction:  
     [LLF - text-style csf](https://bbs.ra2diy.com/forum.php?mod=viewthread&tid=24441&extra=page%3D1), introducing a solution called **LLF Format Text**. This simplifies the CSF file editing process, allowing the game to directly support LLF format files and eliminating the need for traditional CSF editors.  
   - Based on this, the community would benefit from a plugin manager (similar to Minecraft’s mod manager or DP) to handle hook conflicts and other issues.  
   
2. **Custom Toggle-Based**:  
   - This would be a promising direction for Phobos. Ideally, users could customize the compilation process, such as providing a compilation script with a visual UI that allows users to select the features they want to compile.  
   - Alternatively, we could precompile features and offer a feature manager similar to Ares’ INJ file functionality (still requiring a visual UI or user-friendly configuration file) to control which features the user needs.  

---

Phobos 依然是目前社区的一个强大动力引擎，推动着社区的发展。然而，由于程序员数量不足且缺乏资金支持，导致了许多问题。  
对于开发者来说，主要困难在于缺乏维护 PR 和修复 Bug 的人力；而对于用户而言，则面临大量无用的、开发者“写着玩”的功能，以及平台上的大量 Bug，导致整体体验非常不稳定。  

我认为拓展引擎的未来方向应该是 **插件式** 或 **自定义开关式**：  
1. **插件式**：每个功能或组织可以开发一个独立的 DLL 文件，一个 Mod 可以调用多个 DLL。  
   - 在此方向上，我们已经做了一些初步尝试：  
     [LLF-文本可编辑csf](https://bbs.ra2diy.com/forum.php?mod=viewthread&tid=24441&extra=page%3D1)，提供了一种名为 **LLF 格式文本** 的解决方案，旨在简化 CSF 文件的编辑过程，让游戏直接支持 LLF 文件格式，摆脱传统 CSF 编辑器的繁琐操作。  
   - 在此基础上，社区需要一个插件管理器（类似于 Minecraft 的 Mod 管理器或者DP）来处理 Hook 冲突等问题。  
   
2. **自定义开关式**：  
   - 这是 Phobos 的一个理想发展方向。最好能够让用户自定义编译内容，例如提供一个带有可视化 UI 的编译脚本，让用户选择需要编译的功能模块。  
   - 另一种实现方式是我们预先编译好功能，并提供一个类似 Ares 的 INJ 文件功能管理器（需要可视化界面或用户友好的配置文件），使用户能够灵活控制需要启用的功能模块。

**Comment by FS-21** [comment]
Regarding financial support why not apply something like cncNet Open-collective initiative ?

Some stagnated features that are very interesting or useful could get a push if other people push for it. 

In the case the feature contributor does it for free & declines the accumulated ammount then the money goes to the mantainer/dev and if the mantainer/dev declines the ammount then that person can push the feature's ammount to another feature he whishes.

**Comment by reedom114514** [comment]
A review from @ststl-s , translated by AI, post by me due to the network issues, the final interpretation is owned by Ststl:

I believe that the main problem is the lack of testers and the slow review process for PullRequests. Many new features, whether in the development build or PR, are eagerly desired by modders. However, due to the prolonged lack of testing, the official version has fallen seriously behind the version they compiled (branches), or even some PRs have never been merged. This leads to the following issues:

First, when a new development version is released, some features may have bugs or be unnecessary for modders, those who are leading to game lag. This discourages modders from testing new versions. Currently, there are fewer and fewer people making RA2 mods, and they have less time. Modders are more inclined to use a compiled dll that is quickly available and stable. Waiting for an official version that is indefinitely delayed is clearly not an acceptable option for them. In the end, they stop following up on tests and simply use an old version of the development build, further reducing the number of testers for new versions and pushing the official release even further away.

Secondly, For developers, such as myself, due to the slow PR review process and insufficient testing, it is often impossible to merge my features into the development branch. However, my friends may require these new features, leading to the creation of custom versions that mix in some new functions. This often involves self-testing and modifications, which are not easily tracked and merged by Git. Typically, after updating my own version, I would manually make changes again to port them to my own phobos branch (and they might not even be correct, some of my PRs have had issues when just simply copying code over, requiring additional adaptation and correction). As time goes on and my academic workload increases, I have even less time to keep these in sync. Eventually, I stop submitting PRs because each one brings complex merging issues between the official version and my custom version, greatly increasing the time cost and leading to a loss of motivation due to the long period without merges (although having someone use the features you've written is certainly a positive psychological boost, even if it has no practical significance).

In the last, although code style is not critical, it affects the speed of the review process. Given time constraints, reviewers prefer to see simple and understandable code rather than spending time trying to understand someone else's naming conventions and guessing the purpose of functions, variables, and hooks. I have had such significant issues with this: my past PRs have been quite arbitrary, and many, including myself, have had to refactor PRs multiple times. It would be best to provide an excellent example, explaining the naming conventions, to make PRs easier to understand.

Okay, so, Issue 1 is difficult to resolve; the reality that must be faced is the continued decline of the YR modding community, and the inevitable reduction in the number of testers. The only thing developers can do is to speed up the PR review process and the release of the official version, even if it's not perfect, as hotfixes can quickly address issues and motivate testers to continue their efforts. However, this again brings up the issue of the time cost for developers, which is limited, and reviewing others' PRs is less satisfying than creating new features and cannot drive more participation, unless developers are motivated to focus on phobos as the core and find more interest and benefit in pushing for the official release of phobos.

Issue 2 is also a result of the slow iteration of the official version and the slow PR review process. Speeding up reviews and the iteration of the official version could at least improve the situation and reduce the time cost of manual merging.

Issue 3 arises from new developers not understanding PR standards well. Adding an example might help to some extent.

Most of these issues are not solvable and can only be alleviated. The time cost for developers and the time cost for modders to test and iterate are fundamentally conflicting points. What that can be done is to reduce waste of time, ensuring that more time is spent on mutual advancement rather than on manual merging and waiting for the next official version.

**Comment by TaranDahl** [comment]
话说，我们是否需要个官方QQ群或者其它什么的？我知道这是个国际项目，但似乎大多数用户来自中国，这有助于获得他们的反馈。中文互联网不能直接访问Discord，Github也经常连不上。
By the way, do we need an official QQ group or something else? I know this is an international project, but it seems that most users are from China. This will help obtain their feedback. In the Chinese Internet, Discord cannot be directly accessed, and GitHub is often disconnected.

**Comment by Aephiex** [comment]
> 话说，我们是否需要个官方QQ群或者其它什么的？我知道这是个国际项目，但似乎大多数用户来自中国，这有助于获得他们的反馈。中文互联网不能直接访问Discord，Github也经常连不上。 By the way, do we need an official QQ group or something else? I know this is an international project, but it seems that most users are from China. This will help obtain their feedback. In the Chinese Internet, Discord cannot be directly accessed, and GitHub is often disconnected.

Why not? Let's do this.
646311740

**Comment by Aephiex** [comment]
> 话说，我们是否需要个官方QQ群或者其它什么的？我知道这是个国际项目，但似乎大多数用户来自中国，这有助于获得他们的反馈。中文互联网不能直接访问Discord，Github也经常连不上。 By the way, do we need an official QQ group or something else? I know this is an international project, but it seems that most users are from China. This will help obtain their feedback. In the Chinese Internet, Discord cannot be directly accessed, and GitHub is often disconnected.

Why not? Let's do this.

**Comment by TaranDahl** [comment]
> > 话说，我们是否需要个官方QQ群或者其它什么的？我知道这是个国际项目，但似乎大多数用户来自中国，这有助于获得他们的反馈。中文互联网不能直接访问Discord，Github也经常连不上。 By the way, do we need an official QQ group or something else? I know this is an international project, but it seems that most users are from China. This will help obtain their feedback. In the Chinese Internet, Discord cannot be directly accessed, and GitHub is often disconnected.
> 
> Why not? Let's do this.
> 
> 630590659
> 
> (Last one was removed because it was reposted to the Phobos-log channel which may lead to potential leaks of private QQ accounts. Sorry for my privacy anxiety. This comment was initially posted without a QQ group number, and later edited to append the number, meaning it will not be reposted by the bot.)

Good. But we still need the support from the developers to make this official. @Metadorius What do you think?

**Comment by Metadorius** [comment]
> Good. But we still need the support from the developers to make this official. @Metadorius What do you think?

I don't mind and I can advertise it, but I am not sure if the group itself should be called official. Official groups are usually curated by the staff of the project, which would be only partially true in this case, unless there's some way to get in the group for outsiders like me. Maybe something like "Phobos QQ group (maintained by: usernames here)" without explicitly stating it's official?

I have a few questions though:
- How does QQ work at all? Is it an app, a website?
- Is there moderation and who will be doing it if yes?
- Is it possible for non-Chinese users to join?
- If yes - is there some sort of automated translation available?

**Comment by Aephiex** [comment]
> I have a few questions though:
> 
> * How does QQ work at all? Is it an app, a website?
> * Is there moderation and who will be doing it if yes?
> * Is it possible for non-Chinese users to join?
> * If yes - is there some sort of automated translation available?

1. QQ is a Chinese equivalant of discord, except an APP is mandatory.
2. The owner of the group can assign a number of moderators and revoke at will. The capacity depends on how many members are there in a group.
3. Yes, it is, you can either download and use the main branch or go for QQ international. The APP is hard to use and it takes long to tame it, I'm not kidding.
4. I don't think so.

**Comment by DeathFishAtEase** [comment]
> I have a few questions though:
> 
> * If yes - is there some sort of automated translation available?


Unfortunately, QQ only offers manual translation, and you will have to right-click on "Translate" for each conversation to get a result.

**Comment by TaranDahl** [comment]
> I don't mind and I can advertise it, but I am not sure if the group itself should be called official. Official groups are usually curated by the staff of the project, which would be only partially true in this case, unless there's some way to get in the group for outsiders like me. Maybe something like "Phobos QQ group (maintained by: usernames here)" without explicitly stating it's official?

**Comment by TaranDahl** [comment]
> I don't mind and I can advertise it, but I am not sure if the group itself should be called official. Official groups are usually curated by the staff of the project, which would be only partially true in this case, unless there's some way to get in the group for outsiders like me. Maybe something like "Phobos QQ group (maintained by: usernames here)" without explicitly stating it's official?

我的设想是最好能由项目的维护者（或者至少是开发者）组织这个群，收集来自中文社区的意见（例如：哪些已有功能需要改进、哪些pr有很多人希望能尽快合并）以及测试成果（例如某个mod的开发者使用包含pr的构建版本进行了测试且表现良好，那么我们可以收集这个信息用于代码审查时的参考），并将其反馈到Discord上。
简单来说就是让这个群在中文社区作为github issue、discussion和discord的替代，因为一部分中国用户使用github和discord有困难。
My idea is that it would be best if a maintainer (or at least a developer) of the project organizes this group to collect opinions from the Chinese community (for example: which existing functions need improvement, which pull requests many people hope can be merged as soon as possible) and test results (for example, if the developer of a mod uses a build version containing a pr for testing and it performs well, then we can collect this information for reference during code review), and feed it back to Discord. 
In simple terms, this group is used as a substitute for GitHub issues, discussions, and Discord in the Chinese community because some Chinese users have difficulties using GitHub and Discord.

**Comment by Metadorius** [comment]
Alright, I'll have to think about the QQ group for a bit more.

Regarding features being too specific: I agree that having Phobos as a playground for only some of the mod developers who are making features custom-made for their mods is, while a somewhat working strategy, not a very good one because it's a bit one-sided. This is why I said we should strive for generalization of features (where reasonably possible without it turning into a complex programming language on it's own).

On other alternatives:
1) Plug-in architecture - I am not sure if this is easily doable in a sensible manner for this project. This also requires quite a bit of effort to do right and not sure if the benefits outweigh the cost. More so, I believe @Xkein is working on a custom scripting solution, so mod-specific non-generic features could be done that way.
2) A program to provide custom builds - I believe this is a big undertaking as well (the project was not built with this in mind and I have no idea how such a program would work, even less so in a user-friendly manner) with less benefit than generalization of features.
3) Togglable custom features - I don't quite understand what benefit that would bring apart from a minor performance increase?  Most of the features are barely affecting the performance if not used. More so, if the feature is niche, then it ideally wouldn't be in Phobos at all.

Meanwhile I'll ask the next question: if we were to pick a new person to help maintain Phobos/YRpp, review PRs, release new versions etc., who do you think is the most fitting out of all the current active contributors (if at all) and why? I have my own opinion on this which I mostly intend to follow soon but I want to hear some more thoughts.

So, what I think so far is:
1) To solve the problem of too specific features (and partly the pull request amount) we need to set up some sort of boundaries/requirements for the level of reusability, generic-ness and impact of contributions (questions like "how can this be used outside of the designed use case", "how many mod authors will benefit from such a feature"), and work out a workflow for taking new contributions in. The overly specific features could stay in the forks, be done as separate projects based off the framework set up by Ares/Phobos or even be done using DynamicPatcher or Xkein's new solution.
2) The addition of (a) new maintainer(s) and/or triage(s) for pull request is required, even if it might lower the code standards, and I think we should be looking into expanding and maybe even restructuring our maintenance crew soon. I have an opinion and reasoning for who to pick on my own but I also want to hear others too.
3) (Extra based on prev. messages) The release schedule and versioning needs to be rethought. Current stable release cycle is horrendous. I am thinking that maybe we should be doing development builds followed by almost as frequent release builds. Why? Because it is frequent to release a development build then do a bunch of hotfixes. I think we could treat devbuilds as "alpha/beta" builds and then after the hotfixes just release another small build that will be a "stable" build. Also restructuring the versioning scheme may allow for more automated releases (auto changelog and build generation).

Please say if I missed anything important, or leave a reaction if you think we're on right track with this line of thinking so far.

**Comment by Metadorius** [comment]
P.S. for point 1: Phobos is a community project, hence why it should at least try to cater to the wide community and not individual mod devs. And I also think we all should make a better use of discussions and issues.

Discussions are great for thinking out the feature design and they have upvotes which allow people to express whether they think this should be taken in and/or leave comments to improve how the feature should be implemented. For some reason though ever since I moved feature suggestions to discussions everyone stopped caring about them and they barely get submitted, viewed or upvoted.

**Comment by TaranDahl** [comment]
> Discussions are great for thinking out the feature design and they have upvotes which allow people to express whether they think this should be taken in and/or leave comments to improve how the feature should be implemented. For some reason though ever since I moved feature suggestions to discussions everyone stopped caring about them and they barely get submitted, viewed or upvoted.

也许你应该在`README.md`里告诉其他人我们正在那里收集意见。
就我个人的印象来说，Discussion中的内容大多是"我想要一个什么什么新功能"，而且很多都不在我的兴趣范围内。当我刚来到这个项目的时候翻过一遍Discussion，然后就没再去那里看过。
Perhaps you should tell others in the `README.md` that we are collecting opinions there.
Personally, most of the content in Discussion is "I want a certain new function", and many of them are not within my area of interest. When I first came to this project, I flipped through Discussion once and then never looked there again.

**Comment by CrimRecya** [comment]
> (Extra based on prev. messages) The release schedule and versioning needs to be rethought. Current stable release cycle is horrendous. I am thinking that maybe we should be doing development builds followed by almost as frequent release builds. Why? Because it is frequent to release a development build then do a bunch of hotfixes. I think we could treat devbuilds as "alpha/beta" builds and then after the hotfixes just release another small build that will be a "stable" build. Also restructuring the versioning scheme may allow for more automated releases (auto changelog and build generation).

I also quite agree with this point.

我也比较认同这一点。

---

> 我推荐CrimRecya👍
> I recommend CrimRecya.👍

I don't think so.
Although I may be enthusiastic about this, I think the more important thing should be ability to become a maintainer. 
Obviously, I am not competent enough at present. I have some experience in the implementation of some functions, but when it comes to deeper, what I can do is usually copying and pasting.

我觉得不太合适。
尽管我可能对此比较热情，但我觉得要成为维护者更重要的是能力。
显然我还能力不足，虽然对于一些功能的实现我有些许经验，但涉及更底层知识的，我通常只会复制粘贴。

**Comment by Aephiex** [comment]
> > 我推荐CrimRecya👍
> > I recommend CrimRecya.👍
> 
> I don't think so. Although I may be enthusiastic about this, I think the more important thing should be ability to become a maintainer. Obviously, I am not competent enough at present. I have some experience in the implementation of some functions, but when it comes to deeper, what I can do is usually copying and pasting.
> 
> 我觉得不太合适。 尽管我可能对此比较热情，但我觉得要成为维护者更重要的是能力。 显然我还能力不足，虽然对于一些功能的实现我有些许经验，但涉及更底层知识的，我通常只会复制粘贴。

I don't think you have to "learn more" to be "qualified" as a maintainer. Listen, you are a dev, whatever a dev wants they eventually find a path to it, whatever a dev doesn't know they will learn it quickly when needed. It's not a matter of your capacity, it's a matter of your willingness to be a maintainer.

**Comment by Metadorius** [comment]
> I don't think so.
> Although I may be enthusiastic about this, I think the more important thing should be ability to become a maintainer.
> Obviously, I am not competent enough at present. I have some experience in the implementation of some functions, but when it comes to deeper, what I can do is usually copying and pasting.

@CrimRecya I think in these words you devalue your experience (compared to what you are really capable of) a lot :)

**Comment by a851903106** [comment]
One very small question I have is whether old PRs that are too stale like those that have been sitting around for more than 6 months will still be reviewed and merged.

**Comment by Metadorius** [comment]
> I have a very minor question as to whether old PRs that are too old (sitting for 6 months or even a year or more) will still be reviewed and merged.

We do want to want to still review (and merge if they're fitting) them. The "PR fridge" as we call it was not intentional.

> However, the wait was so long that I almost lost my enthusiasm to submit them.
That's sad but understandable. Hope we'll be able to solve the situation with PRs and maintainers so there would be a point to submit the PRs again. Thanks for understanding the situation and sticking around with us.

**Comment by Starkku** [comment]
Some of what I am about to write has probably been stated above already but I'll just throw in my two cents regardless.

First I'm gonna reiterate something I have mentioned at least once before, in some words or other.

# Number of open PR's alone is not indicative of anything of value

The open PR's can be anything, ranging from useful and desirable features implemented in a way that matches the project's scope and quality standards to random stuff scribbled together in a haste or out of scope features. The number is not an useful metric to follow here.

The point is that if our metric is gonna be number of merged PR's instead of having more focused objectives to fulfill then we might as well give up right now. It is not worth it at that point, the end result will serve no one as it will be a bloated mess unsuitable for modder's needs and only serves to keep the non-existent maintainers busy.

As for the solution or better approach, I unfortunately do not have surefire answers to that but I believe that...

# Phobos needs a client - a group of people that serve as the users of the product that dictate what features need to be implemented

This cannot be every random person piping in the feature requests. It needs to be way more focused. It cannot be driven by the thoughts and ideas of the developers themselves especially if they are not end-users of Phobos. The people comprising this group are most likely going to have to be **active users of Phobos in existing projects**, e.g modders with established projects. A case could be made for including more fresh meat as well but that would likely require further investigation or discussion.

My point is that if we're gonna have to be more selective about what features get work done on and implemented, and make no mistake, this is definitely the case currently, it has to be done with the needs of the end-users placed first, wishes of the developers second. Of course there needs to be a discussion or process to this and not just follow every whim of the client group, that wouldn't work either. I don't know about the practical implementation of such process either, unfortunately, but I believe it is something that warrants further discussion and action.

Some things I think would be important considerations when thinking of any future features or implementations regarding Phobos. I am not going to mention anything regarding code style and documentation or lack thereof as they are fairly obvious things and have been stated before.

- **End-user applications:** Already effectively stated above, but if it is a feature that no one has practical use for then it is not worth anyone's time.
- **Flexibility vs. performance:** This is a tough one, some features in Phobos already steer towards the wrong side of this (and I am guilty for some of it myself). This old-ass engine has a massive scalability issue and often adding extra flexibility to features can end up aggravating the problem. It is something that has to be considered on case-by-case basis and it should be one of the primary concerns of the developer responsible as performance degradation should be avoided at all costs.
- **Code complexity, reusability, refactoring etc:** KISS. Keep it simple, stupid! Investigate if what you're about to implement has not already been done before, f.ex does game already have a function that does exactly what you want to do instead or reinventing the wheel, this one I feel like is particularly important has repeatedly popped up in the past. Do not copy paste, split reused or reusable bits of code into functions, inlined if necessary for performance (or just let compiler decide, I think they are pretty smart about this stuff). Boilerplate can sometimes be reduced by using inheritance, but I am aware that some systems already in place such as AE and shields do not follow this principle even though they probably should.

**Comment by TaranDahl** [comment]
> Number of open PR's alone is not indicative of anything of value

不敢苟同。
如果你认为部分pr因为某些原因不应该被接收，那你可以制定一个容易判断的要求，并发公告或者其它什么来告诉贡献者他们需要满足这个要求，并且关掉那些不符合要求的pr。
而现实是，目前大部分pr处于被搁置的状态。没有人审查并给出意见，或者少数审查意见都已被修改完毕。开发者不知道下一步该做什么，因为没人来告诉他们。维护pr需要时间和精力，有大量开启状态的pr就意味着在浪费那些人的时间和精力。而持续性地缺少反馈会让那些开发者感到被忽视，这会消耗他们的热情，最终他们离开并建一个自己的分支自娱自乐。
我说，既然Phobos的定位是“a community engine extension project”，那么请给予你的“community”以足够的重视。至少应该让贡献者知道“我们需要这个”或者“我们不需要这个”，而不是持续无目的无止境的等待。我觉得这不难做到，对吧？
I cannot agree. 
If you think that some pull requests should not be accepted for certain reasons, then you can formulate an easily judged requirement, issue an announcement or do something else to tell contributors what requirements they need to meet, and close those pull requests that do not meet the requirements. 
But the reality is that most pull requests are currently in a suspended state. No one reviews and gives opinions, or the few review opinions have been revised. Developers don't know what to do next because no one tells them. Maintaining pull requests requires time and energy. Having a large number of open pull requests means wasting the time and energy of those people. And the continuous lack of feedback will make those developers feel ignored, which will consume their enthusiasm. Eventually, they will leave and create their own branch for self-entertainment. 
I say, since the positioning of Phobos is "a community engine extension project", then please pay enough attention to your "community". At least let contributors know "we need this" or "we don't need this", instead of continuous aimless and endless waiting. I think this is not difficult to do, right?

**Comment by Starkku** [comment]
That is the charitable and probably intended interpretation of a 'community project' but you have already outlined yourself why it does not work out in reality, there's simply not enough manpower to manage a project in that fashion. Hell, that's why this discussion even exists in the first place. There is no system in place to currently triage features effectively which is very much needed and my suggestion was one approach to that problem. Not necessarily the only one, though.

**Comment by TaranDahl** [comment]
> That is the most charitable yet probably intended interpretation of a 'community project' 

我没想到我们对于'community project'的解读会存在分歧。也许我对此有一些过高的期望？
I didn't expect that there would be a disagreement in our interpretation of 'community project'. Maybe I had some overly high expectations for this?

**Comment by TaranDahl** [comment]
对于设定精确目标的想法，我想说说我的担忧。存在这么一种可能就是你设定了一个有用且有意义的目标但没人去做。一个例子就是Kerbiter之前多次提过的EventHandler，这个东西确实有用而且有意义，但就是没有足够的人手来把它实现。如果我们不能把想法实现的话那么想法再好也意义不大。
当然我也注意到了你是这个项目开发端扛把子，如果你打算自己兜底的话那就不存在任何问题了。
Regarding the idea of setting precise goals, I would like to express my concerns. There is a possibility that you set a useful and meaningful goal but no one works on it. An example is the EventHandler that Kerbiter has mentioned many times before. This thing is indeed useful and meaningful, but there are not enough people to implement it. If we cannot implement our ideas, then no matter how good the ideas are, they are of little significance. 
Of course, I have also noticed that you are the carrier on the development side of this project. If you plan to finish the goal by yourself on the occasion mentioned above, then there will be no problem.

**Comment by Aephiex** [comment]
Don't think we need a long term goal. We are not a real team, but a bunch of developers happen to be here doing their own stuff on a same open source project, no one is giving anyone else an order, everyone is working on whatever they are interested. To give a goal to such a group of people you have to first make them a team, to make them a team you have to first make them decide who commands whom. I can't say on behalf of the most, but on behalf of my self, I will refuse to follow anyone's command and will only do whatever I want, thus I'm not a team joiner here.

**Comment by TaranDahl** [comment]
> If yes - is there some sort of automated translation available?
我成功建设了一个bot用作中英互译。所以这个问题算是解决了。
I successfully built a bot for Chinese-English translation. So this problem is considered solved.

**Comment by Metadorius** [comment]
> I successfully built a bot for Chinese-English translation. So this problem is considered solved.

Wow, that's great! The new problem, though, is that I can't register in QQ from my IP address, even with the help of @Aephiex and @CrimRecya, since I don't have a Chinese phone number. (I appreciate it, thanks!)

I've also invited @CrimRecya into Phobos team as one of the steps to alleviate the issues we're facing, congratulations on the promotion @CrimRecya! I haven't updated the README list of Phobos team members yet cause I am planning on a slight restructure/clarification on team structure, contribution types etc. in docs, partly in order to make more people able to help with maintenance. I plan to submit a draft PR with my ideas so we have something to discuss further.

Partly I am inspired by this: https://bevyengine.org/learn/contribute/project-information/bevy-organization/

**Comment by Metadorius** [comment]
This is ot ready for a pull request, but since I was mainly occupied with familiarising myself with the Chinese part of the community, I figured I still should post a draft of what I am slowly coming up with and hear some feedback. Names are temporary.

```md

### Phobos maintenance crew structure

- **Leads** - primary decision makers, define where the project is headed
- **Ring 1 maintainers** - assigned to controversial PRs, make releases, ...
- **Ring 2 maintainers** - assigned to less controversial PRs, ...
- **Triages** - help triage/label PRs, issues, discussions, help with communication with the community
### Phobos contribution types

- Reconnection error (desync) and crash (fatal error) fixes
  - Should be the highest priority ideally
  - Default controversy level is low. Can be reviewed/committed directly by ring 2 maintainers.
- Phobos bugfixes
  - Default controversy level is low. Can be reviewed/committed directly by ring 2 maintainers.
- Vanilla bugfixes
  - Default controversy level is low. Can be reviewed/committed directly by ring 2 maintainers.
- Unhardcodings/customizations - contributions that make something customizable through the INI or other way by the modder usually.
  - Default controversy level is low. Can be reviewed/committed directly by ring 2 maintainers.
- New features (maybe we must further break down this?)
  - Must be triaged on reusability by other modders (how?)
  - Default controversy level is medium. Can be reviewed/committed directly by ring 1 maintainers.

What can make any PR more controversial and requiring a higher level maintainer's attention:
- Modifying/breaking previous (or vanilla) behavior
- Requiring migration
- Mixing contribution types
- Not fitting into any of the standard types or not being like anything that was submitted earlier to Phobos
- Amount of code a lot bigger than typical for the contribution type

The amount of attention (as in amount of people who should review) that is needed scales with:
- amount of code
- amount of API surface, as in how many INI (rule of thumb: check docs size)
```

This should allow more people maintain Phobos and look for fixes etc., as this assumes more easier contribution types that can be handled by more people.

This doesn't yet address how to gauge whether a contribution fits in Phobos and should be merged, if not - what could be done to fix that, and how to objectively measure all that. I was thinking of using discussions to get upvotes and discussion going first, but this seems to work more for features that are not yet submitted.

**Comment by KenosisM** [comment]
If you guys are looking for "what features should be made", why not take a look at a certain mature, flexible old engine, the SC2 galaxy engine? The basic concepts like "search area" "launch projectile" mostly inspired Kratos. I'd say most of its functions are the try of copying a Galaxy engine on YR. Of course, writing INI under Kratos is a little complicated, but a good low-code node editor will save most of that pain, just like SC2 modding on its map editor is just filling charts/linking modules.


it includes literally "everything" - model, sounds, etc. Only green nodes are game logic. 

My answer is making a trigger-event system. While on shooting, on start moving, on health percent, on hit, on death, on creation, on having certain behavior (AE, buff), on X frames later, on expiration of event, on event stack count, etc, Do search area (search area would also link to events: on every acquired target, do......), launch projectile, play animation, teleport, spawn units, launch sw, attach/remove a set of effects, etc.

**Comment by TaranDahl** [comment]
> - Must be triaged on reusability by other modders (how?)

可能并非所有PR都有必要询问其它modder。
从上面的讨论来看，人们对于未开启的功能主要的担忧在于是否会影响性能。从个人的实践经验来说，目前Phobos在不启动任何特性的情况下对性能的影响是可忽略的。实际上在这个issue之外我从未听说过任何人对Phobos的性能进行过抱怨。
Perhaps not all pull requests need to ask other modders. 
From the above discussion, people's main concern about unactivated functions is whether it will affect performance. From personal practical experience, currently the impact of Phobos on performance when no features are activated is negligible. In fact, outside of this issue, I have never heard anyone complain about the performance of Phobos.

**Comment by TaranDahl** [comment]
Anyway, having a document is always much better than having none. Glad you have done this.

**Comment by TaranDahl** [comment]
> This doesn't yet address how to gauge whether a contribution fits in Phobos and should be merged, if not - what could be done to fix that, and how to objectively measure all that. I was thinking of using discussions to get upvotes and discussion going first, but this seems to work more for features that are not yet submitted.

我觉得你需要对Phobos给出一个准确的定位，这有助于我们进一步讨论该合并什么的问题。
我个人来说，Phobos与其它引擎扩展相比独特的地方在于它开源、底层没有严重缺陷、而且对于贡献没有很多限制。我不太希望它有很强制的方向，至少对贡献者来说是这样。
I think you need to give an accurate positioning of Phobos. This will help us further discuss what should be merged. 
Personally, the unique thing about Phobos compared to other engine extensions is that it is open source, has no serious defects at the bottom layer, and there are not many restrictions on contributions. I don't want it to have a very mandatory direction, at least for contributors.

**Comment by TaranDahl** [comment]
> This is ot ready for a pull request, but since I was mainly occupied with familiarising myself with the Chinese part of the community, I figured I still should post a draft of what I am slowly coming up with and hear some feedback. Names are temporary.

有一个问题你没有关注到：谁来测试PR？PR被合并前需要进行怎样程度的测试？
There is a problem you haven't noticed: Who will test the pull request? To what extent does testing need to be done before a pull request is merged?

**Comment by ahsy666** [comment]
I think it's time to make a special version with most useful functions. It's been too long after 0.3.0.1 released. Mod makers really need a new powerful Phobos with few bug. And, please keep giving patch to make this version as stable as possible, until you gonna make another new stable version with new functions.

**Comment by DeathFishAtEase** [comment]
> I think it's time to make a special version with most useful functions. It's been too long after 0.3.0.1 released. Mod makers really need a new powerful Phobos with few bug. And this time, please keep giving patch to make this version as stable as possible, until you gonna make another new version with new functions.

Welcome to test [releasev0.4](https://github.com/Phobos-developers/Phobos/pull/1459) and [Build#46](https://github.com/Phobos-developers/Phobos/releases). If you are interested in the new features, you can go to the [Pull Request](https://github.com/Phobos-developers/Phobos/pulls) page to help test the PR.
